include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h

```diff
@@ -1520,12 +1683,48 @@ namespace internal
                 else
                   cell_coarse->get_mg_dof_indices(local_dof_indices);
 
+                std::vector<types::global_dof_index> local_dof_indices_lex(
+                  local_dof_indices.size());
+
+                for (unsigned int i = 0; i < local_dof_indices.size(); ++i)
+                  if (lexicographic_mappings.size() > 0)
+                    local_dof_indices_lex[i] =
+                      local_dof_indices[lexicographic_mappings[fe_index_coarse]
+                                                              [i]];
```

Comment:
The compile bug looks like some of those errors we had when too many lambda functions were nested; I don't see that as the case here and in particular not against something that changed here, but it might still somehow overflow here. As I assume this does not happen on newer still maintained gcc versions, there is no point in submitting a bug report on a reduced test case, so we just work around it. I think this ICE is a sign to reduce the number of nested statements and break things out into separate (named) functions, as this `reinit` function has turned into a spaghetti function with too much straight-line code and too many indirections into separate functions by lambdas.

The two things I would recommend here would be to break out `process_cells` into a separate free-standing function similarly to the `loop_overactive_or_level_cells`, and to try to use similar code for the two functions to be used on the two lambda functions passed into it (they only appear to differ in the data fields that get filled as far as I can see, but I admit I got lost in checking what exactly gets duplicated).